### PR TITLE
Fixes in Tensorboard web app

### PR DESCRIPTION
**Pull request by kimwnasptd**
These are some small fixes required for the Tensorboards web app to be able to work under a prefix and inside a container, like jupyter-web-app.

/cc @elikatsis 
/cc @StefanoFioravanzo 

The last remaining part then will be to create the manifests for this app cc @yanniszark

### CODE REVIEW

File: components/crud-web-apps/tensorboards/frontend/package.json [hunk 1]
```diff
@@ -4,7 +4,7 @@
   "scripts": {
     "ng": "ng",
     "start": "npm run copyLibAssets && ng serve --base-href /tensorboard/ --deploy-url /tensorboard/",
-    "build": "npm run copyLibAssets && ng build --prod --base-href /tensorboard/ --deploy-url /tensorboard/static/",
+    "build": "npm run copyLibAssets && ng build --prod --base-href /tensorboards/ --deploy-url static/",
```
Comment (DavidSpek) [inline components/crud-web-apps/tensorboards/frontend/package.json:7]: In the line above the base-href is `/tensorboard/` instead of `/tensorboards/` which it is here.

File: components/crud-web-apps/tensorboards/Dockerfile [hunk 2]
```diff
@@ -48,4 +48,4 @@ COPY ./tensorboards/backend/Makefile .
 
 COPY --from=frontend /src/dist/ /src/app/static/
 
-ENTRYPOINT make run
+ENTRYPOINT gunicorn -w 3 --bind 0.0.0.0:5000 --access-logfile - entrypoint:app
```
Comment (nrchakradhar) [inline components/crud-web-apps/tensorboards/Dockerfile:51]: Will this have any incompatibilities for IPv6?

Comment (DavidSpek) [reply components/crud-web-apps/tensorboards/Dockerfile:51]: @nrchakradhar Good point. These same entrypoint is being used by the Jupyter-Web-App. Do you have suggestions how to make this IPv6 compatible?

Comment (nrchakradhar) [reply components/crud-web-apps/tensorboards/Dockerfile:51]: I do not know anything about gunicorn, but from its documentation, the following may work:
https://docs.gunicorn.org/en/stable/settings.html
Multiple addresses can be bound. ex.:

$ gunicorn -b 127.0.0.1:8000 -b [::1]:8000 test:app
will bind the test:app application on localhost both on ipv6 and ipv4 interfaces.

I found an interesting observation here(its quite old though)
https://github.com/benoitc/gunicorn/issues/1138#issuecomment-158358666

Can we make this configurable from manifests so that IPv6 support is provided. There are few issues created already for IPv6 support: #5615 #5605

File: components/crud-web-apps/common/backend/kubeflow/kubeflow/crud_backend/helpers.py [hunk 3]
```diff
@@ -24,7 +24,7 @@ def get_prefixed_index_html():
     with open(os.path.join(static_dir, "index.html"), "r") as f:
         index_html = f.read()
         index_prefixed = re.sub(
-            r"\<base href=\".*\"\>", '<base href="%s">' % prefix, index_html,
+            r"\<base href=\".*\".*\>", '<base href="%s">' % prefix, index_html,
```
Comment (elikatsis) [inline components/crud-web-apps/common/backend/kubeflow/kubeflow/crud_backend/helpers.py:27]: What's the issue we are trying to fix here? Do we need to match **anything** that follows the `href` in the `<base>` tag?

File: components/crud-web-apps/tensorboards/Dockerfile (hunk 2 above)
Comment (elikatsis) [inline components/crud-web-apps/tensorboards/Dockerfile:51]: Can we use the `/bin/bash -c` notation here so that
1. we explicitly use `bash` instead of `sh` (https://docs.docker.com/engine/reference/builder/#shell-form-entrypoint-example), and
2. it is uniform with what we do for other web apps (JWA https://github.com/kubeflow/kubeflow/blob/28af7716bc22f25158ddde05af58d53f4fafc939/components/crud-web-apps/jupyter/Dockerfile#L50)

File: components/crud-web-apps/tensorboards/backend/app/utils.py [hunk 4]
```diff
@@ -17,19 +16,12 @@ def parse_tensorboard(tensorboard):
...
         "namespace": tensorboard["metadata"]["namespace"],
         "logspath": tensorboard["spec"]["logspath"],
-        "age": helpers.get_age(tensorboard),
+        "age": tensorboard["metadata"]["creationTimestamp"],
         "status": status.create_status(phase, message, "")
     }
 
     return parsed_tensorboard
 
-def getPVCName(pvc):
```
Comment (elikatsis) [inline components/crud-web-apps/tensorboards/backend/app/utils.py:26]: Could you split the changes of this commit in 2?
1. date component related changes
2. PVC name related changes

When this PR is squashed & merged we will still maintain a description of this change in the final commit message

**Review by elikatsis — COMMENTED** [review]
Thanks @kimwnasptd !

I left some comments here and there

File: components/crud-web-apps/tensorboards/frontend/package.json (hunk 1 above)
Comment (kimwnasptd) [reply components/crud-web-apps/tensorboards/frontend/package.json:7]: I'm not sure I understand what the question or suggested change is here. Could you rephrase this?

Comment (DavidSpek) [reply components/crud-web-apps/tensorboards/frontend/package.json:7]: There is an inconsistency in the usage of `--base-href /tensorboard/ --deploy-url /tensorboard/"` on Line 6 and `--base-href /tensorboards/` on Line 7.

File: components/crud-web-apps/common/backend/kubeflow/kubeflow/crud_backend/helpers.py (hunk 3 above)
Comment (kimwnasptd) [reply components/crud-web-apps/common/backend/kubeflow/kubeflow/crud_backend/helpers.py:27]: @elikatsis I wanted to catch the use case where the `base` tag could look like this, which includes a space as well
```html
<base href="/" >
```

File: components/crud-web-apps/tensorboards/Dockerfile (hunk 2 above)
Comment (kimwnasptd) [reply components/crud-web-apps/tensorboards/Dockerfile:51]: Makes sense, will add a commit for this

File: components/crud-web-apps/tensorboards/backend/app/utils.py (hunk 4 above)
Comment (kimwnasptd) [reply components/crud-web-apps/tensorboards/backend/app/utils.py:26]: makes sense

File: components/crud-web-apps/tensorboards/frontend/package.json (hunk 1 above)
Comment (kimwnasptd) [reply components/crud-web-apps/tensorboards/frontend/package.json:7]: I see what you mean. The `start` script should be completely removed, since the users should run the UI locally with `build:watch` for now.

I'll push a commit to remove this script

File: components/crud-web-apps/tensorboards/Dockerfile (hunk 2 above)
Comment (kimwnasptd) [reply components/crud-web-apps/tensorboards/Dockerfile:51]: @nrchakradhar thank you for your comments. I'm planning on tackling the IPv6 support for all the web apps after 1.3.

The first step is to bind the IPv6 equivalent addresses as well, as you've pointed out in your links. But I want to do a good deep dive on this and test some edge cases for which I don't have the cycles right now.

File: components/crud-web-apps/tensorboards/frontend/package.json (hunk 1 above)
Comment (kimwnasptd) [reply components/crud-web-apps/tensorboards/frontend/package.json:6]: Added a commit that removes the unused npm script

File: components/crud-web-apps/common/backend/kubeflow/kubeflow/crud_backend/helpers.py (hunk 3 above)
Comment (elikatsis) [reply components/crud-web-apps/common/backend/kubeflow/kubeflow/crud_backend/helpers.py:27]: Then I'd probably suggest using a regex like `r"\<base href=\".*\"\s?\>"` just to make things a bit more strict. Would this work?

Comment (kimwnasptd) [reply components/crud-web-apps/common/backend/kubeflow/kubeflow/crud_backend/helpers.py:27]: Yeap, I'll use that instead

Comment (kimwnasptd) [reply components/crud-web-apps/common/backend/kubeflow/kubeflow/crud_backend/helpers.py:27]: @elikatsis in the end it's not just the white spaces. We have two cases:
1. `<base href="/tensorboards/">`
2. `<base href="/" />`

There's also an extra `/` character in the end that I missed. So if you are OK I'll stick to the current catch-all rule I had which captures this as well, since we will be complicating the regex more to handle the `/`.

File: components/crud-web-apps/tensorboards/backend/app/utils.py (hunk 4 above)
Comment (kimwnasptd) [reply components/crud-web-apps/tensorboards/backend/app/utils.py:26]: force pushed the changed commits. Resolving this one

File: components/crud-web-apps/tensorboards/Dockerfile (hunk 2 above)
Comment (kimwnasptd) [reply components/crud-web-apps/tensorboards/Dockerfile]: pushed a commit, will resolve this

File: components/crud-web-apps/common/backend/kubeflow/kubeflow/crud_backend/helpers.py (hunk 3 above)
Comment (elikatsis) [reply components/crud-web-apps/common/backend/kubeflow/kubeflow/crud_backend/helpers.py:27]: Sure let's move on with this as is and come back to it in the future if there's any reason to do so.

**Comment by kimwnasptd** [comment]
@elikatsis I've also pushed my commits that address the review comments. The tests also pass, so feel free to merge if you don't have any further suggestions for the time

**Comment by elikatsis** [comment]
Thanks @kimwnasptd !

/lgtm
/approve

**Comment by DavidSpek** [comment]
@kimwnasptd  @PatrickXYS Just want to highlight here that merging this PR has also triggered a postsubmit for the Volumes and Jupyter Web Apps. It seems like prow is not taking the diff from the PR into account, but rather the diff compared to Master. This way, if a PR is not up to date with the latest commit in master, it will also trigger pre- and postsubmit jobs for any other components that have since changed in the master branch. 
So in this case, I think the images created with the postsubmit build might actually be a regression of the latest changes in the Volumes and Jupyter Web Apps.
